src/hooks/useMaciVote.ts

```diff
@@ -0,0 +1,94 @@
+import { publishBatch, type Signer } from "maci-cli/sdk";
+import { useCallback, useState } from "react";
+
+import { config } from "~/config";
+import { useEthersSigner } from "./useEthersSigner";
+import { useSession } from "next-auth/react";
+import { useMaciSignup } from "./useMaciSignup";
+import { useMaciPoll } from "./useMaciPoll";
+
+export interface IVoteArgs {
+  voteOptionIndex: bigint;
+  newVoteWeight: bigint;
+}
+
+export interface IUseMaciVoteData {
+  isLoading: boolean;
+  error?: string;
+  onVote: (
+    args: IVoteArgs[],
+    onError: () => Promise<void>,
+    onSuccess: () => Promise<void>,
+  ) => Promise<void>;
+}
+
+export const useMaciVote = (nonce = 0): IUseMaciVoteData => {
+  const signer = useEthersSigner();
+  const { data } = useSession();
+  const { stateIndex } = useMaciSignup();
+  const { pollData } = useMaciPoll();
+  const [isLoading, setIsLoading] = useState(false);
+  const [error, setError] = useState<string>();
+
+  const onVote = useCallback(
+    async (
+      votes: IVoteArgs[],
+      onError: () => Promise<void>,
+      onSuccess: () => Promise<void>,
+    ) => {
+      if (!signer || !data || !stateIndex || !pollData) {
+        return;
+      }
+
+      if (!votes.length) {
+        await onError();
```

Comment:
why first onError() then setError?

Reply:
Actually, it doesn't matter. With this react method it will be set anyway.